controller/orphan_controller.go

```diff
@@ -0,0 +1,371 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+)
+
+type OrphanController struct {
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+	// use as the OwnerID of the controller
+	controllerID string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	ds *datastore.DataStore
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewOrphanController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string) *OrphanController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	oc := &OrphanController{
+		baseController: newBaseController("longhorn-orphan", logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-orphan-controller"}),
+	}
+
+	ds.OrphanInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    oc.enqueueOrphan,
+		UpdateFunc: func(old, cur interface{}) { oc.enqueueOrphan(cur) },
+		DeleteFunc: oc.enqueueOrphan,
+	})
+	oc.cacheSyncs = append(oc.cacheSyncs, ds.OrphanInformer.HasSynced)
+
+	ds.NodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Use `AddEventHandlerWithResyncPeriod` to disable resync for non-primary resource watching, the same as other controllers.